### PR TITLE
Supress empty buffer warning when multiple sources are searched

### LINKED ISSUE #175: Silence the "org-ql: No headings in buffer: %s" ?
This is a minor thing, but the "org-ql: No headings in buffer: %s" message on empty buffers is a bit annoying when using ql non-interactively. (I have an agenda buffer filled by an ical subscription, it's often empty …) Could that message perhaps only be sent if called interactively, or perhaps there could be some variable we can set? Or should we just inhibit-message when calling org-ql (feels a bit heavy-handed)?

### PR DESCRIPTION
**Pull request by colonelpanic8**
Fixes #175

### CODE REVIEW

**Comment by alphapapa** [comment]
For an option like this, which will probably be set only rarely, it would probably be better to use a `defcustom` than to add another argument to various functions.

Also, please be careful not to introduce formatting, indentation, or whitespace changes in a patch.

Finally, you should make a feature branch on your clone and send the PR from that.  Sending PRs from your own master branch is usually a recipe for trouble.

**Comment by colonelpanic8** [comment]
> Also, please be careful not to introduce formatting, indentation, or whitespace changes in a patch.

I mean theres bad trailing whitespace there. You really don't want to fix that?

> For an option like this, which will probably be set only rarely, it would probably be better to use a `defcustom` than to add another argument to various functions.

Doesn't this defeat the whole purpose of having that there in the first place?

> Finally, you should make a feature branch on your clone and send the PR from that. Sending PRs from your own master branch is usually a recipe for trouble.

Hmmm I think I can do with my fork what I want to.

I actually did this intentially, because I've set up straight to point to my fork because I don't want to wait for this change, because its extremely annoying to get this message.

**Comment by alphapapa** [comment]
> > Also, please be careful not to introduce formatting, indentation, or whitespace changes in a patch.
> 
> I mean theres bad trailing whitespace there. You really don't want to fix that?

Trailing whitespace is a trivial matter.  It's not important or urgent to fix.  It's more important to keep a patch or PR minimal so as to be easy to review.  As well, keeping unrelated changes out of a commit means that it can later be reverted if necessary without also reverting the unrelated fixes.

So if you would like to contribute whitespace fixes, please do so in a separate PR.

> > For an option like this, which will probably be set only rarely, it would probably be better to use a `defcustom` than to add another argument to various functions.
> 
> Doesn't this defeat the whole purpose of having that there in the first place?

I don't understand what you mean.  What I would like to do is avoid making the code more complicated in 4 places and having to pass an additional argument to a function for a very minor matter.

> > Finally, you should make a feature branch on your clone and send the PR from that. Sending PRs from your own master branch is usually a recipe for trouble.
> 
> Hmmm I think I can do with my fork what I want to.

Indeed, you can reject the shared wisdom of fellow community developers if you like.  I can also choose to reject PRs from other users' master branches, because it makes my job harder (e.g. if you change your master branch again, the commit history becomes more complicated, and I can't force-push changes to this PR before merging because it's your master branch rather than a feature branch, etc).  This is basic git "hygiene" and standard practice in the community.

> I actually did this intentially, because I've set up straight to point to my fork because I don't want to wait for this change, because its extremely annoying to get this message.

That's not relevant.  Your PR should be from a branch on your fork that is based on this fork's master branch.  Then you can merge the patch into your own fork's master branch as you like, without affecting anything or anyone else.

**Comment by colonelpanic8** [comment]
...cool man. have a nice day.

**Comment by alphapapa** [comment]
@colonelpanic8 FYI, this kind of behavior is not common in the Emacs community.  Many users contribute bug reports, feature requests, and patches, and they take feedback politely, coordinating changes until it's ready to merge.  It's not necessary to be rude and passive-aggressive.  If everyone acted that way, we wouldn't even be here.  

It's not an insult to ask for changes to be made to a proposed patch.  When I've contributed patches to Emacs itself, I've often gone through a few rounds of feedback and changes, even ones that only mattered to please the maintainer's preferences.  It's a matter of being respectful to those who contribute their time to maintain something for us to benefit from using.

**Comment by colonelpanic8** [comment]
> rude and passive-aggressive.

how was I rude? Can you quote a comment for me?

You have to realize that stuff like this is super patronizing:

> Finally, you should make a feature branch on your clone and send the PR from that. Sending PRs from your own master branch is usually a recipe for trouble.

What my branch is called doesn't affect your repo at all, and I'm really free to do whatever I want with my fork. I generally do things like this, but in this case I decided not to, because I figured this would be merged quickly and I did the change in like 2 seconds because I wanted the annoying noise to stop.

Ultimately I closed the PR because I'm just not invested enough to spend the time to do it exactly the way that you want to do it, and I already have a solution that works for me.

You should probably fix the issue in my opinion, but if you don't its not the end of the world, I'll just change my straight recipe to point at my fork.

> It's not an insult to ask for changes to be made to a proposed patch.

No its not, but I honestly find your taste on some of these subjects to be really bad, and I found all of the commentary on the way I was doing things to be grating. For example, keeping changes small and self contained and easy to review is a good principle, for sure, but you're taking it a little far when you nitpick one whitespace fix.

Part of being a good maintainer is understanding how to balance asking things of people with maintaining certain standards around how things are done.


ALSO:

The fact that you EVEN have the message there in the first place is honestly pretty weird in my opinion, but the way I fixed it went out of my way to sort of "try to respect the principle" of what you were trying to accomplish. I'll agree that what you suggested is a better approach (although I'd argue that just removing the message would be the best approach), but I sort of figured that the type of person that would even want the message there in the first place would prefer something like this to just having a variable to suppress things.

**Comment by ivanperez-keera** [comment]
I was going to let this go to avoid stirring things, but I think it's important that I speak up.

I agree with @alphapapa that your response sounded passive aggressive. I thought the same when you wrote "...cool man. Have a nice day." and then just closed the issue.

Perhaps you didn't intend it that way. But, to at least two people, it came across that way.

> What my branch is called doesn't affect your repo at all, and I'm really free to do whatever I want with my fork.

I also ask people to send PRs from a branch other than master (it's a rule I set in my repos and completely unrelated to org-ql or even the emacs community). When people don't do it that way, it kind of forces me to create a branch locally on my machine, pull to it from the user's repo, merge it locally, and then push my new master.

You are free to make changes to master in your branch, but when you do send a PR, that changes things on `org-ql`s side, and it also affects you if you intend to become a longer term contributor (for the reasons that @alphapapa explained). This last bit may not be your intention, but it's perfectly ok for @alphapapa to explain that and request it.

>  I honestly find your taste on some of these subjects to be really bad.

Maybe, but I don't recall @alphapapa asking for your advice on that, and PR certainly isn't the place to offer such commentary, nor is that the tone.

Before this PR became a more heated discussion, I was going to post a comment saying that I wish more people applied such rules in their own repos. Adding whitespace changes to an unrelated PR makes things harder to review, makes changes harder to undo (if needed). It's just bad SW engineering, but the community tends to do it out of what in my opinion is probably just laziness.

I didn't post such a message because I thought you could interpret it in a way that was not intended and it might offend you.

It's refreshing to see open-source software projects that adopt a clean and rigorous approach and say "put whitespace changes elsewhere if they are unrelated".

You are free to disagree with some rules, but I don't think it is appropriate that you come to this PR to try to educate @alphapapa on how he should maintain his own project.

TL;DR: Please, just try to be more considerate next time, understand that @alphapapa is volunteering his own time, and try to be more appreciative of his effort creating and maintaining this library even, or *especially*, if you'd do things differently.

**Comment by colonelpanic8** [comment]
> You are free to make changes to master in your fork, but when you do send a PR, that changes things on `org-ql`s side, and it also affects you if you intend to become a longer term contributor (for the reasons that @alphapapa explained). This last bit may not be your intention, but it's perfectly ok for @alphapapa to explain that and request it.

branches are literally JUST POINTERS. When you merge, you are merging the commit hash, not the branch.

> When people don't do it that way, it kind of forces me to create a branch locally on my machine, pull to it from the user's repo, merge it locally, and then push my new master.

It really sounds like you don't actually know how git works. There is literally no reason for you to do that.

**Comment by ivanperez-keera** [comment]
@colonelpanic8 Thank you for your kind words. Have a great evening.

**Comment by colonelpanic8** [comment]
> @colonelpanic8 Thank you for your kind words. Have a great evening.

dude, read what you wrote here:

> When people don't do it that way, it kind of forces me to create a branch locally on my machine, pull to it from the user's repo, merge it locally, and then push my new master.

and tell me how it makes any sense. I'm not even trying to be mean, but you can literally just admit that you are wrong. The process you just described there literally has NO effect on the end result.

The best argument there is is that I might potentially make a mistake and use my master branch for something else by accident. but literally there is no impact on the actual result of a merge depending on the alias that you pass to git when you run the merge command.

if a branch first-branch and another branch second-branch point to the same commit,

```
git merge first-branch
```

and 

```
git merge second-branch
```
produce exactly the same result.

Go ahead, try it on your machine.

**Comment by ivanperez-keera** [comment]
@colonelpanic8 "you can literally just admit that you are wrong"

Once again, this tone is absolutely inappropriate. I'm not going to participate in any discussion that you approach that way.

I have better things to do.

Now, leave me alone.

**Comment by alphapapa** [comment]
@ivanperez-keera I appreciate your kind words and thoughtful defense.

@colonelpanic8 I can't add much beyond what @ivanperez-keera and I have already said.  

There are sound technical reasons to not send pull requests based on your fork's master branch.  If you won't believe us, then a quick googling provides [this](https://blog.jasonmeridth.com/posts/do-not-issue-pull-requests-from-your-master-branch/) as the first result, a blog article from 11 years ago, explaining it in more detail.  In the several years I've been maintaining projects on GitHub and accepting contributions to them, I've experienced this problem numerous times, and now I have learned to avoid it by not accepting PRs from a fork's master branch.  

If you don't understand that, or if you simply don't care because it doesn't seem to affect you, then it's really inappropriate for you to accuse another developer of not knowing how git works, and to accuse me of being patronizing for explaining it.

> When you merge, you are merging the commit hash, not the branch.  It really sounds like you don't actually know how git works. 

I'm far from an expert on git, but I've been using it for a number of years, and it sounds like you need to do some review, especially with regard to how the commit graph works, how branches work, and how merging works.  You may be conflating merging with applying patches or cherry-picking.

By the same token that you are entitled to do what you want with your projects, I am so entitled with regard to mine.  I choose to keep aesthetic changes separate from functional ones for what are in my opinion good technical reasons, which I already explained.  

If you're unwilling to follow my projects' standards in your proposed contributions, that's fine--I haven't asked you to contribute, and it's your time to use as you please.  But it's not fine for you to respond to polite feedback and patient explanations by being rude, passive-aggressive, and outright insulting myself and another contributor.

I ask that any further contributions you make on my projects, whether discussion or code, be respectful.  There's no reason to be unfriendly; we're all on the same side here, trying to make Emacs and its related projects better.

**Comment by colonelpanic8** [comment]
> I'm far from an expert on git, but I've been using it for a number of years, and it sounds like you need to do some review, especially with regard to how the commit graph works, how branches work, and how merging works. You may be conflating merging with applying patches or cherry-picking.

bro:
https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell#:~:text=A%20branch%20in%20Git%20is,branch%20pointer%20moves%20forward%20automatically.

"A branch in Git is simply a lightweight movable pointer to one of these commits. The default branch name in Git is master. As you start making commits, you’re given a master branch that points to the last commit you made. Every time you commit, the master branch pointer moves forward automatically."

The only sense in which a merge commit refers to the child is through the commit hash. Branches literally do not exist from the perspective of git history.

It's incredibly condescending for you to say 'it sounds like you need to do some review", when you are in fact wrong. Merge commits NEVER point to branches. The only sense in which a merge can even have ANYTHING to do with a branch is when github automatically includes something in the merge commits comment, but this has no effect on the git history itself.

**Comment by colonelpanic8** [comment]
> I ask that any further contributions you make on my projects, whether discussion or code, be respectful. There's no reason to be unfriendly; we're all on the same side here, trying to make Emacs and its related projects better.

Sure, I totally agree with that sentiment. I really feel like I was not actually rude to you if you read my first comment.

I'm sorry we got off on the wrong foot, but I think you should also think about how your initial comments came off, and realize that you were absolutely super condescending.

**Comment by colonelpanic8** [comment]
> especially with regard to how the commit graph works,

Please enlighten me as to how anything I've said is wrong about the commit graph. Cite 1 source that explains how branches feature in the commit graph.

You have to understand that when you write something like this:

> I'm far from an expert on git, but I've been using it for a number of years, and it sounds like you need to do some review, especially with regard to how the commit graph works, how branches work, and how merging works. You may be conflating merging with applying patches or cherry-picking.

you better be DAMN sure that the person you're talking to is actually wrong.

Can you actually not see how condescending that is?

**Comment by alphapapa** [comment]
@colonelpanic8 I've lost track of how many times you've made accusations of being condescending and patronizing.  That's enough.
